sha1: Roll iterations inside transform() into loop

andyhhp opened the pull request:
Use 5-round unrolled loops to keep the a..e variables consistent with
before.  The change in key schedule at round 16 is awkward, so leave
that block as it was previously.

Passing a non-constant into M() requires more brackets to produce the
correct logic, so take the opportunity to simplify it, by pulling out
the logic look up in x mod 16, and drop the temporary variable
entirely.

64:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-2458 (-2458)
Function                                     old     new   delta
transform                                   4433    1975   -2458
Total: Before=55933, After=53475, chg -4.39%

32:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-2140 (-2140)
Function                                     old     new   delta
transform                                   4048    1908   -2140
Total: Before=27030, After=24890, chg -7.92%

lto.64:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-2458 (-2458)
Function                                     old     new   delta
transform                                   4433    1975   -2458
Total: Before=55466, After=53008, chg -4.43%

lto.32:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-2140 (-2140)
Function                                     old     new   delta
transform                                   4048    1908   -2140
Total: Before=26298, After=24158, chg -8.14%

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>

rossphilipson reviewed (approved):
Loops look like they are correctly doing what the removed code was doing. My approval still stands after the force update. I think this PR is good to go.

Reviewed-by: Ross Philipson <ross.philipson@oraclec.com>

andyhhp committed:
sha1: Roll iterations inside transform() into loop

Use 5-round unrolled loops to keep the {a..e} variables consistent with
before.  The change in key schedule at round 16 is awkward, so leave
that block as it was previously.

64:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-2458 (-2458)
Function                                     old     new   delta
transform                                   4433    1975   -2458
Total: Before=55933, After=53475, chg -4.39%

32:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-2140 (-2140)
Function                                     old     new   delta
transform                                   4048    1908   -2140
Total: Before=27030, After=24890, chg -7.92%

lto.64:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-2458 (-2458)
Function                                     old     new   delta
transform                                   4433    1975   -2458
Total: Before=55466, After=53008, chg -4.43%

lto.32:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-2140 (-2140)
Function                                     old     new   delta
transform                                   4048    1908   -2140
Total: Before=26298, After=24158, chg -8.14%

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
Reviewed-by: Ross Philipson <ross.philipson@oracle.com>
Reviewed-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Tested-by: Krystian Hebel <krystian.hebel@3mdeb.com>

andyhhp committed:
sha256: Fix sha256_blend() to be idempotent

Rename BLEND_OP() to sha256_blend() so it is clearer to spot in disassembles.

The index to blend can be derived from the loop index, and shouldn't be stored
in a static variable.  This allows the compiler better flexibility with
instruction scheduling, and makes the algorithm thread-safe.

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
Reviewed-by: Ross Philipson <ross.philipson@oracle.com>
Reviewed-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Tested-by: Krystian Hebel <krystian.hebel@3mdeb.com>

andyhhp committed:
sha256: Fold sha256_transform() some more

64:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-733 (-733)
Function                                     old     new   delta
sha256_update                               3035    2302    -733
Total: Before=53397, After=52664, chg -1.37%

32:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-536 (-536)
Function                                     old     new   delta
sha256_update                               3021    2485    -536
Total: Before=24859, After=24323, chg -2.16%

lto.64:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-733 (-733)
Function                                     old     new   delta
sha256_update                               3035    2302    -733
Total: Before=52921, After=52188, chg -1.39%

lto.32:
add/remove: 0/0 grow/shrink: 0/1 up/down: 0/-536 (-536)
Function                                     old     new   delta
sha256_update                               3021    2485    -536
Total: Before=24079, After=23543, chg -2.23%

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
Reviewed-by: Ross Philipson <ross.philipson@oracle.com>
Reviewed-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Tested-by: Krystian Hebel <krystian.hebel@3mdeb.com>

andyhhp committed:
sha256: Drop big endian types

Update sha256_{transform,update,final}() to consume void pointers, which
removes some casts.  Drop LOAD_OP() entirely, as it adds more complexity than
it saves.  Remove unused includes.

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
Reviewed-by: Ross Philipson <ross.philipson@oracle.com>
Reviewed-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Tested-by: Krystian Hebel <krystian.hebel@3mdeb.com>

andyhhp committed:
sha1: Extra sha1_blend() out of sha1_transform()

Fixing sha256_blend() showed that the blend operation itself is expensive to
inline.  Offer the compiler the option to out-of-line it, which it does take.

64:
add/remove: 1/0 grow/shrink: 1/1 up/down: 44/-916 (-872)
Function                                     old     new   delta
sha1_blend                                     -      43     +43
sha1_write                                   225     226      +1
sha1_transform                              4433    3517    -916
Total: Before=55388, After=54516, chg -1.57%

32:
add/remove: 1/0 grow/shrink: 0/1 up/down: 48/-296 (-248)
Function                                     old     new   delta
sha1_blend                                     -      48     +48
sha1_transform                              4039    3743    -296
Total: Before=26442, After=26194, chg -0.94%

lto.64:
add/remove: 1/0 grow/shrink: 1/1 up/down: 44/-916 (-872)
Function                                     old     new   delta
sha1_blend                                     -      43     +43
sha1_write                                   227     228      +1
sha1_transform                              4433    3517    -916
Total: Before=55127, After=54255, chg -1.58%

lto.32:
add/remove: 1/0 grow/shrink: 0/1 up/down: 48/-296 (-248)
Function                                     old     new   delta
sha1_blend                                     -      48     +48
sha1_transform                              4039    3743    -296
Total: Before=25983, After=25735, chg -0.95%

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
Reviewed-by: Ross Philipson <ross.philipson@oracle.com>
Reviewed-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Tested-by: Krystian Hebel <krystian.hebel@3mdeb.com>

andyhhp committed:
sha1: Rename transform() to sha1_transform()

To avoid it being ambiguous with sha256_transform()

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
Reviewed-by: Ross Philipson <ross.philipson@oracle.com>
Reviewed-by: Krystian Hebel <krystian.hebel@3mdeb.com>
Tested-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel reviewed (approved):
Both SHA1 and SHA256 return the same values as before

Reviewed-by: Krystian Hebel krystian.hebel@3mdeb.com
Tested-by: Krystian Hebel krystian.hebel@3mdeb.com
